api/hasura/actions/create_circle.ts

```diff
@@ -0,0 +1,173 @@
+  const insertProfiles = {
+    objects: [{ address: address }, { address: coordinapeAddress }],
+    on_conflict: {
+      constraint: profiles_constraint.profiles_address_key,
+      update_columns: [profiles_update_column.address],
+    },
+  };
+  const insertUsers = {
+    data: [
+      {
+        name: object.user_name,
+        address: address,
+        role: 1,
+      },
+      {
+        name: 'Coordinape',
+        address: coordinapeAddress,
+        role: 2,
+        non_receiver: false,
+        fixed_non_receiver: false,
+        starting_tokens: 0,
+        non_giver: true,
+        give_token_remaining: 0,
+        bio: 'Coordinape is the platform you’re using right now! We currently offer our service for free and invite people to allocate to us from within your circles. All funds received go towards funding the team and our operations.',
+      },
+    ],
+  };
+  const circleReturn = {
+    id: true,
+    name: true,
+    protocol_id: true,
+    team_sel_text: true,
+    alloc_text: true,
+    vouching: true,
+    min_vouches: true,
+    nomination_days_limit: true,
+    vouching_text: true,
+    logo: true,
+    default_opt_in: true,
+    team_selection: true,
+    only_giver_vouch: true,
+    auto_opt_out: true,
+  };
+  try {
+    // attach new circle to existing organisation
+    if (object.protocol_id) {
+      //check if user is an circle admin within organisation
+      const { profiles } = await gql.q('query')({
+        profiles: [
+          {
+            where: {
+              address: { _eq: address },
+              users: {
+                role: { _eq: 1 },
+                circle: { protocol_id: { _eq: object.protocol_id } },
+              },
+            },
+          },
+          { id: true },
+        ],
+      });
+      try {
+        z.array(
+          z.object({
+            id: z.number(),
+          })
+        )
+          .nonempty({
+            message:
+              'Address is not an admin of any circles under this protocol',
+          })
+          .parse(profiles);
+      } catch (err) {
+        if (err instanceof z.ZodError) {
+          return res.status(422).json({
+            extensions: err.issues,
+            message: 'Invalid input',
+            code: '422',
+          });
+        }
+      }
+
+      const { insert_circles_one } = await gql.q('mutation')({
+        insert_circles_one: [
+          {
+            object: {
+              name: object.circle_name,
+              protocol_id: object.protocol_id,
+              users: insertUsers,
+            },
+          },
+          circleReturn,
+        ],
+        insert_profiles: [
+          insertProfiles,
+          {
+            returning: {
+              id: true,
+            },
+          },
+        ],
+      });
+      return res.status(200).json(insert_circles_one);
+    } else {
+      // create a new organisation
+      const { insert_organizations_one } = await gql.q('mutation')({
```

Comment:
We will want to move these into a library file, like `src/lib/gql/GqlHasuraAdmin.ts`

Reply:
Done